docs/exceptions.md

```diff
@@ -0,0 +1,266 @@
+document.
+
+### Regular error
+
+Consider a function like `readFile`. It is entirely possible that the path
+passed to it may point to a file that does not exist, and the programmer can't
+do anything to prevent this. Existence of a file is not a property of our code,
+it's the property of the outside world, and we have to deal with all possible
+scenarios.
+
+Another example is parsing user input. We might expect the user to enter a
+number, but we have to consider the case that the user enters something else.
+
+In cases when the erroneous scenarios are out of our control, we consider
+these to be regular errors.
+
+### Identifying error class
+
+Can the execution of the error codepath be excluded by code review and static
+verification (more precise types?)
+
+* Yes: it is a programmer mistake
+* No: it is a regular error
+
+## Is the code 100% pure or potentially impure?
+
+### Impure code
+
+The code is considered impure when it's an `IO` action, a function that returns
+an `IO` action, or similar. The code can be _potentially_ impure when it is
+written in an abstract monad that can be instantiated to `IO`.
+
+Definitely impure:
+
+```
+readFile :: FilePath -> IO String
+```
+
+Potentially impure:
+
+```
+lookupThrow :: MonadThrow m => Map k v -> k -> m v
+```
+
+### Pure code
+
+Pure code is one that does not use `IO` operations. For example:
+
+```
+lookupMaybe :: Map k v -> k -> Maybe v
+```
+
+Notice that while it is conceivable that we have a `Map` of `IO` actions, and
+instantiate `lookupMaybe` to `Map k (IO ()) -> k -> Maybe (IO ())`, the
+instantiation does not inspect/use these `IO` actions, so it's not considered
+potentially impure.
+
+### Identifying code purity
+
+Is there an instantiation of type parameters that would mean that there are
+`IO` actions inspected/used in the code?
+
+* Yes: it is potentially impure code
+* No: it is 100% pure code
+
+## Error handling practices
+
+### Pure code, programmer mistakes
+
+Do *not*:
+
+* do this often, as if it's okay
+* use `undefined`
+* use non-exhaustive pattern matching
+
+Do:
+
+* try to use types to avoid the need in the first place
+* comment extensively (invariants and precondition, reasoning)
+* use the `error` function (or `impureThrow` with custom exception)
+* use the `HasCallStack` feature
+
+Before writing code that allows programmer mistakes, consult with colleagues how
+it would be possible to prevent them statically. There might be tricks that
+you're not aware of. Always make this trade-off conciously and responsibly.
+
+(On the other hand, if static guarantees require GADTs or type families, perhaps
+it's better to not overcomplicate code. Use your judgement.)
+
+DISCUSSION: Should we create a synonym `bug = impureThrow` in Universum? This
+would make the intention more clear.
+
+### Pure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* use `MonadFail`
+* return `Either Text`
```

Comment:
Why not? Sometimes we have `someFoo = runExceptT someFooDo` because every usecase would anyway run `runExceptT`  on the `someFooDo`. Or do you mean that returning `Text` error is bad? It's a reasonable tradeoff, i guess.

Reply:
This is about `Text`. Define proper ADTs for your errors.

Reply:
I can't see why is it always reasonable.

Reply:
I answered in the neighboring thread where you made the same comment.